chore(testing): initial Kurtosis integration for exploratory migration

IreneBa26 opened the pull request:
## Motivation

This PR explores an initial approach to **migrating** Lodestar's testing framework from **Docker-based** execution to **Kurtosis-managed** Ethereum network simulations. It proposes an abstraction layer that could **preserve compatibility** with existing test assertions while shifting the underlying infrastructure.
The core logic is maintaining the **exact same test assertions** while swapping the implementation from Docker containers to Kurtosis services.

**This PR is opened for feedback and team discussion about the architectural approach before proceeding with the complete implementation.**

This work is part of the **Ethereum Protocol Fellowship - Cohort 6**. 
More details, including weekly updates on implementation and proposals, can be found in this [documentation](https://hackmd.io/@IB26).

## Description

### **Proposed Solution**

Implement a Kurtosis SDK-based runner that provides:

- **Enclave Management**: Isolated execution environments with automatic cleanup
- **Service Orchestration**: Declarative network configuration via YAML
- **API Abstraction**: Consistent interfaces regardless of underlying execution method
- **Backward Compatibility**: Existing test assertions continue to work unchanged

### **Technical Implementation**

**1. Core Architecture (runner/)**

- **KurtosisSDKRunner**: Implements IRunner interface, inspired by Docker Runner pattern
- Manages enclave lifecycle (create, start, stop, destroy)
- Deploys services using ethpandaops/ethereum-package Starlark package
- Maps Kurtosis services to NodeService abstraction layer
- **kurtosisTypes.ts**: Type definitions for network configuration and service metadata
- **loadKurtosisConfig.ts**: YAML configuration loader for test scenarios

**2. Interface Layer (simulation/interfaces-kurtosis.ts)**

- **Extended IRunner**: Modified to accept KurtosisNetworkConfig instead of JobOptions[]
- **Service Mapping**: KurtosisServicesMap provides typed access to Kurtosis-managed services
- **Backward Compatibility**: Maintains existing NodePair, BeaconNode, ExecutionNode interfaces
- **Kurtosis-Specific Extensions**: New interfaces like BeaconNodeKurtosis for future use

**3. Simulation Integration (simulation/simulation-kurtosis.ts)**

- **initWithKurtosisConfig()**: Static factory method for Kurtosis-based simulations
- **Service Conversion**: Maps Kurtosis NodeService objects to Crucible-compatible NodePair structures
- **API Client Integration**: Creates Lodestar API clients using Kurtosis service endpoints
- **Lifecycle Management**: Integrates with existing simulation tracker and epoch clock
- **createNodePairsFromKurtosis()**: Converts Kurtosis services to Crucible NodePair[]
- **createBeaconNodeFromKurtosis()**: Creates BeaconNode with Kurtosis service context

**4. Test Infrastructure (test/)**

- **multi-fork.yml**: Example network configuration demonstrating multi-client setups
- **runner-test.ts**: Test suite for KurtosisSDKRunner
- Service creation and lifecycle validation
- Port mapping and network topology verification
- Service role inference and metadata extraction

### **Data Flow Architecture**

YAML Config → KurtosisNetworkConfig → KurtosisSDKRunner → KurtosisServicesMap → NodeService[] → NodePair[] → Test Assertions

Example of a possible workflow:
```mermaid
flowchart TD
    A["multiFork.test.ts"] --> B["Simulation.initWithKurtosisConfig()"]
    B --> C["Simulation constructor"]
    C --> D["new Kurtosis Runner()"]
    
    D --> E["multiFork.yml"]
    E --> F{"Configuration Defines"}
    F --> G["EL/CL Types & Counts"]
    F --> H["Additional Services<br/>e.g., dora"]
    F --> I["Fork Epochs<br/>Altair, Bellatrix, Capella,<br/>Deneb, etc."]
    
    D --> K["Loads YAML"]
    K --> L["Launches Enclave"]
    L --> M{"Returns"}
    M --> N["Beacon API URLs"]
    M --> O["EL/CL RPC Endpoints"]
    M --> P["Service Identifiers<br/>cl-1, el-2, etc."]
    
    N --> Q["Simulation"]
    O --> Q
    P --> Q
    Q --> R["Passes to Assertions"]
    R --> S["No API Change Required"]
```

### **Future Work**

**If team feedback is positive on the architectural approach:**

- Investigate if and how Kurtosis services can be fully replicated into **compatible** NodePair objects
- Research compatibility between Kurtosis service endpoints and existing Lodestar API interfaces (IRunner, BeaconNode, etc)
- Attempt to **fully replicate** one existing simulation test using Kurtosis instead of Docker

**Note**: The current approach only explains the architectural strategy. Actual NodePair replication from Kurtosis services is **not yet implemented** and requires further investigation to determine feasibility and approach

### **Feedback Request**

This PR is opened **specifically for team feedback and discussion on**:

1. **Runner & Architecture**: Is the current Kurtosis runner implementation and overall substitution strategy the right approach?
2. **Interface & Types**: Do the proposed Kurtosis types/abstractions fit well, or should they be rethought?
3. **Migration Path**: Does the step-by-step transition plan look correct, or should it be structured differently?

IreneBa26 committed:
chore(kurtosis): add initial runner and mock simulation scaffolding
**Motivation**

Exploratory work for replacing Docker-based simulation infra
with Kurtosis. Goal is to validate feasibility and document integration
points with the current test framework

**Description**

- Add KurtosisSDKRunner to wrap enclave lifecycle and expose services (working)
- Add runner-test.ts to run the KurtosisSDKRunner and print service info from Kurtosis (working)
- Add simulation-kurtosis.ts and interfaces-kurtosis.ts as mock scaffolding
  → contain errors / unused params, not usable, shared only for feedback
- Add kurtosisTypes.ts and helper loadKurtosisConfig.ts for typing and YAML parsing

**Notes**

Only KurtosisSDKRunner and runner-test.ts are functional
Other files are placeholders for structure review and not production-ready

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=8296) <br/>All committers have signed the CLA.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/runner/kurtosisSDKRunner.ts:
```diff
@@ -0,0 +1,131 @@
+   *
+   * enclaveName: Default name for the Kurtosis enclave - can be overridden
+   */
+  constructor(enclaveName = "crucible-enclave") {
```
Make the name required, so the enclave name already be unique with the simulation running. And don't mixup the services among different simulations.

nazarhussain replied:
Make the name required, so the enclave name already be unique with the simulation running. And don't mixup the services among different simulations.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/simulation/interfaces-kurtosis.ts:
```diff
@@ -0,0 +1,497 @@
+import {ChildProcess} from "node:child_process";
```
Seems this file contains a lot of duplicate interfaces from other `interface.ts` file. Keep it consistent and add new interfaces in the `interface.ts`.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/runner/loadKurtosisConfig.ts:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  const raw = await fs.readFile(fullPath, "utf8");
+  return parse(raw) as KurtosisNetworkConfig;
```
Is there a way to validate the kurtosis config at this point? may be some helper function in their SDK.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/multi-fork.yml:
```diff
@@ -0,0 +1,15 @@
+participants:
```
We also need to update relevant `multiFork.ts` file so instead of inline declerations of nodes, it loads up the Yaml file.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/runner/kurtosisSDKRunner.ts:
```diff
@@ -0,0 +1,131 @@
+        id: serviceName,
+        serviceContext,
+        beaconApiUrl: ports.get("http") ? `http://localhost:${ports.get("http")?.number}` : undefined,
+        roles: this.inferRoles(serviceName),
```
Can there be a service with multiple roles? I think not.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/runner/loadKurtosisConfig.ts:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  const raw = await fs.readFile(fullPath, "utf8");
+  return parse(raw) as KurtosisNetworkConfig;
```
Is there a way to validate the kurtosis config at this point? may be some helper function in their SDK.

IreneBa26 replied:
The `create()` function calls `runStarlarkRemotePackageBlocking()`, which in turn invokes `sanity_check()` and `input_parser()` from the ethereum-package. 

These already provide an initial validation of the Kurtosis config. Is this sufficient?

If sufficient but should occur earlier, one possible adjustment could be to move the validation (happening in _simulation-kurtosis.ts_) to run right after `await env.runner.start(`sim-${id}`);` and before `const services = await env.runner.create(kurtosisConfig);`, so that it happens before `create()` rather than inside it

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/simulation/interfaces-kurtosis.ts:
```diff
@@ -0,0 +1,497 @@
+  Nethermind = "execution-nethermind",
+}
+
+export enum ExecutionStartMode {
```
We may delete this one now. Not needed ny more.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/runner-test.ts:
```diff
@@ -0,0 +1,258 @@
+/**
```
Remove this file from this PR and keep it locally for your testing.

IreneBa26 committed:
refactor(kurtosis): address PR feedback

- Update NodeService to use single role instead of multiple boolean flags
- Change beaconApiUrl to generic apiUrl for better abstraction
- Enclave name changed to be unique with the simulation running
- Addressed kurtosis config validation
- Added new interfaces in the interface.ts
- Add TODO comment indicating future merge of simulation-kurtosis into simulation
- Updated multiFork.test.ts wiith the new kurtosis configuration
- Removed runner-test.ts from the PR
- Improved multi-fork.yml parameters to be similar to the multifork.test.ts config

nazarhussain commented on .gitignore:
```diff
@@ -42,6 +42,9 @@ packages/beacon-node/mainnet_pubkeys.csv
 packages/api/oapi-schemas
 test-logs/
 
+# Kurtosis test files (keep locally, don't commit)
+packages/cli/test/utils/crucible/kurtosis/test/runner-test.ts
+
```
We would remove this line when this PR get ready.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/simulation/simulation-kurtosis.ts:
```diff
@@ -396,7 +395,7 @@ export class Simulation {
 
   //❌ REMOVE - Docker-specific
-  async createNodePair<B extends BeaconClient, V extends ValidatorClient, E extends ExecutionClient>({
+  /*async createNodePair<B extends BeaconClient, V extends ValidatorClient, E extends ExecutionClient>({
```
You can remove the code which is not needed from this file anyway.

IreneBa26 replied:
As previously suggested for `interfaces-kurtosis.ts` to keep it consistent and append new code, should `simulation-kurtosis.ts` be unified into `simulation.ts` and just rely on `interfaces.ts` + `simulation.ts`, removing `simulation-kurtosis.ts` and `interface.ts.` entirely?

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/multi-fork.yml:
```diff
@@ -1,14 +1,25 @@
   - el_type: geth
     cl_type: lodestar
-    count: 4
+    #vc_type: lodestar
+    #cl_extra_params: ["--builder.seelction=default"]
+  - el_type: geth
+    cl_type: lodestar
+    #vc_type: lodestar
+    #cl_extra_params: ["--builder.seelction=default"]
+  - el_type: geth
+    cl_type: lodestar
+    #vc_type: lodestar
+    #cl_extra_params: ["--builder.seelction=executiononly"]
+  - el_type: geth
+    cl_type: lodestar
+    #vc_type: lodestar
+    #cl_extra_params: ["--builder.seelction=default"]
```
Should not we just create a network config yaml file for Kurtosis and then load it up here in TS?

nazarhussain replied:
Was confused with some other file. Ignore this message.

IreneBa26 committed:
refactor(kurtosis): network config improvements

- Update simulation-kurtosis.ts
- Removed runner-test.ts from the PR
- Removed Docker Runner elements
- Aligned multi-fork.yml parameters to the multifork.test.ts config, as in unstable branch

IreneBa26 committed:
refactor(kurtosis): NodePair[] integration

**Motivation**
- Extend the Kurtosis-based simulation to fully reconstruct NodePair[] objects by dynamically creating Beacon, Execution, and Validator nodes from discovered services
- Ensure that each node type (CL/EL/VC) is correctly initialized with consistent network parameters, improving parity with the original Docker-based workflow

**Description**
- Added createBeaconNodeFromKurtosis, createExecutionNodeFromKurtosis, and createValidatorNodeFromKurtosis to instantiate Lodestar, Geth, and Lighthouse services directly from Kurtosis
- Introduced automatic role and index inference so nodes can be accurately matched across layers (cl-1 with el-1)
- Integrated flexible port discovery (getPort, httpUrlFrom) to handle diverse client naming conventions across services
- Refactored the simulation lifecycle to use the beacon genesis and health endpoints as readiness gates instead of generating local genesis files

**Notes**
- The multi-fork simulation can now be executed with the regular test command (yarn test:sim:multifork), using the Kurtosis-backed runner end-to-end
- syncing.ts adaptation to Kurtosis is still in progress

nazarhussain commented on packages/cli/test/sim/deneb.test.ts:
```diff
@@ -19,36 +18,13 @@ const {estimatedTimeoutMs, forkConfig} = defineSimTestConfig({
-      remote: true,
-    },
-  ]
+  "deneb.yml"
```
I would suggest to keep `deneb.test.ts` and `deneb.yml` in the same directory. So as the other test files. Would be much easier to navigate and check the tests.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/deneb.yml:
```diff
@@ -0,0 +1,35 @@
+  - el_type: geth
+    cl_type: lodestar
+    vc_type: lodestar
+    count: 1
```
What is this count? is it necessary to specify in all network config files? What's the default value if we skip it here?

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/deneb.yml:
```diff
@@ -0,0 +1,35 @@
+participants:
```
We have only 2 nodes in the `deneb.test.ts`. Let's try to mimic the same environment here and have only two participants.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/multi-fork.yml:
```diff
@@ -0,0 +1,42 @@
+    
+keymanager_enabled: true
+network_params:
+  prefunded_accounts: '{"0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "100ETH"}, "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa": {"balance": "100ETH"}}'
```
Can we use it in YAML format rather JSON string?

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/multi-fork.yml:
```diff
@@ -0,0 +1,42 @@
+  prefunded_accounts: '{"0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b": {"balance": "100ETH"}, "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa": {"balance": "100ETH"}}'
+  num_validator_keys_per_node: 32 #emulate keysCount
+  preset: minimal
+  altair_fork_epoch: 0
+  bellatrix_fork_epoch: 0
+  capella_fork_epoch: 0
+  deneb_fork_epoch: 0
```
These forks does not match with the ones we had in the `multiFork.test.ts` file.

nazarhussain commented on packages/cli/test/utils/crucible/kurtosis/test/network-basic.yml:
```diff
@@ -0,0 +1,9 @@
+participants:
```
If we are not using this file, then remove it.

nazarhussain commented on packages/cli/test/utils/crucible/utils/network.ts:
```diff
@@ -22,27 +22,35 @@ export async function connectNewNode(newNode: NodePair, nodes: NodePair[]): Prom
+    const src = (await node.api.node.getNetworkIdentity()).value();
+    if (src?.peerId === targetPeerId) continue;
+
+    // 4) In Kurtosis, do NOT rewrite to 127.0.0.1
```
👍🏼

nazarhussain commented on packages/cli/test/utils/crucible/interfaces.ts:
```diff
@@ -136,7 +135,7 @@ export interface ExecutionGenesisOptions<E extends ExecutionClient = ExecutionCl
   extends ExecutionGenesisOptions<E>,
     GeneratorOptions {
-  mode: ExecutionStartMode;
+  //mode: ExecutionStartMode; //✅ REMOVED - ExecutionStartMode not needed anymore
```
Yes remove that line completely.
